initial setup

schuttsm opened the pull request:
Api endpoints tested:
* auth/api/v1/users/me (get)
* projects/business/<business_id>/projects (get)
* timetracking/business/<business_id>/time_entries (get & post)
* timetracking/business/<business_id>/time_entries/time_entry_id (delete)
* auth/oauth/token (post)

I setup the travis build to run nightly with emails to you & me if it fails (or on change when it succeeds).

Ready for review: @prust

Commit:
setup selenium tests on saucelab to acquire the authorization code.  use that to get a refresh token & run tests.js with mocha

Commit:
add time entries create, list & delete api calls. adjust email settings in travis

servant-42 commented on .travis.yml:
```diff
@@ -0,0 +1,21 @@
+language: node_js
+node_js:
+- stable
```
@schuttsm: It'd be nice to [test this on Node 4, 6 and 8](https://docs.travis-ci.com/user/languages/javascript-with-nodejs/#Specifying-Node.js-versions)

servant-42 commented on freshbooks.js:
```diff
@@ -70,17 +84,28 @@ FreshBooks.prototype._request = function(path, data, callback) {
+  var method = 'get';
+  if (data) {
+      headers['content-type'] = 'application/json';
+      method = data.method;
```
@schuttsm: won't overloading `data` to carry the method as well as the JSON to be posted result in an extra `method` property being added to all records posted to the API? It may be better to have `method` as a separate, optional argument, or to have `this.get()`/`del()`/`post()`/etc methods.

servant-42 commented on freshbooks.js:
```diff
@@ -70,17 +84,28 @@ FreshBooks.prototype._request = function(path, data, callback) {
+    json: data,
+    method: method
+  }, function(err, res, body) {
+    if (!err && res.statusCode == 204 && data.method == 'delete')
```
@schuttsm: All 2xx codes are success, so it'd be a little simpler & more robust to just check for that instead of specific response codes for specific methods.

servant-42 commented on freshbooks.js:
```diff
@@ -70,17 +84,28 @@ FreshBooks.prototype._request = function(path, data, callback) {
 
     try {
-      body = JSON.parse(body)
+      if (typeof body == 'string')
```
@schuttsm: totally optional, but `_.isString()` is a little less verbose than `typeof`.

The reason for the branching might be a little clearer if we do `JSON.parse()` on GET requests and not on non-GET requests? (but maybe not as robust? Your choice...)

servant-42 commented on package.json:
```diff
@@ -24,12 +25,15 @@
-    "chai": "^4.1.2",
-    "dotenv": "^4.0.0",
-    "mocha": "^5.0.0"
+    "chai": "4.1.2",
```
@schuttsm: `chai` is duped between the lists, not sure why it needs to be in the top list, but if it does then it doesn't need to be here. I'm also not sure why `wdio` needs to be in the top list...

servant-42 replied:
@schuttsm: We should be able to switch to the recommended practice of using the `^` prefix for all dependencies here, since this a library and we're using `package-lock.json` in the apps that use it.

servant-42 commented on setup.js:
```diff
@@ -0,0 +1,38 @@
+    const openTabs = browser.getTabIds();
+    const currentTab = browser.getCurrentTabId();
+    browser.switchTab(openTabs[0]);
+    browser.close(openTabs[1]);
```
@schuttsm: This surprised me -- what's in the 2nd tab?

servant-42 commented on tests.js:
```diff
@@ -1,44 +1,103 @@
+
+        project_id = projects[0].id;
+        chai.assert.ok(projects.length);
+        chai.assert.ok(projects[0].id);
```
@schuttsm: This is a great start, but I think we should verify the presence & the type of every property we're using (or we should stop using it if we don't need to):

In `bc-proxy.js`, `loadProjectsForBiz()` I think that also includes:

* `body.meta.pages` and `body.meta.page` (i think we're expecting both to be numbers)
* `proj.client_id`
* `proj.title`
* `proj.updated_at`
* `proj.active` (looks like we're expecting a boolean?)

Also, `getActiveBusinesses()` in `api-utils.js` depends on some `subscription_statuses` and `business_memberships` data (I think originally we didn't have this, but got errors from the API when we attempted to pull data from businesses whose subscription has lapsed). This data is being saved to the Firebase `people/{psn_id}/service_data/freshbooks` by the `/fbooks-callback` route in `bcsyncer/index.js`. Also, we have *got* to consolidate as much of the FreshBooks API code into this lib as possible, I'm going cross-eyed trying to find it across all the different containers it's scattered in. I'm sorry about the mess, most of it made some kind of sense at the time (though I admit some of it was pragmatic).

prust reviewed (approved):
@schuttsm: This is great, I'm really impressed! There is more to go (see below) in terms of verifying the API returns all the data we're depending on, and also including all the FreshBooks API logic that is in our various service, but this is a great start and it's perhaps best to do additional work in subsequent pull requests.

Commit:
remove selenium setup, manually generate authorization code & then use refresh tokens for each test run after that.  cleanup dependencies

Commit:
remove selenium test step from travis.yml

Commit:
finish removing saucelabs

Commit:
add code logging for debugging purposes

Commit:
fix travis tests

Commit:
add logging

Commit:
save refresh token or authorization code in firebase for next test run

Commit:
run travis for node 4, 6 & 8

Commit:
cleanup freshbooks.js

Commit:
add guard for no refresh token

Commit:
improve logging when refresh_token is invalid

Commit:
cleanup test code

Commit:
fix tests

Commit:
expand test coverage

Commit:
remove selenium setup file

Commit:
remove wdio.conf file

schuttsm commented:
@prust - okay, all the review changes have been implemented.  I removed all the selenium tests & just store the code or refresh token in firebase.  I've had the refresh_token get out of sync every so often, all I do is regenerate the authorization code in freshbooks & then reinsert it into firebase (in the env_data table there is a single object - if it has a code, it uses that to generate a refresh token, if it has a refresh token it uses that, then saves the resulting refresh token for the next call)

ready for another round of review

servant-42 commented on package.json:
```diff
@@ -23,13 +23,13 @@
   "homepage": "https://github.com/CSNW/node-new-freshbooks#readme",
-  "dependencies": {
-    "request": "^2.83.0",
-    "underscore": "^1.8.3"
```
@schuttsm: i think both these deps need to be included here instead of below in devDependencies

servant-42 commented on .gitignore:
```diff
@@ -56,4 +56,6 @@ typings/
-
+refreshToken.txt
+errorShots
+code.txt
```
@schuttsm: these 3 lines aren't necessary anymore, are they?

servant-42 commented on freshbooks.js:
```diff
@@ -10,84 +10,99 @@ function FreshBooks(access_token, refresh_token) {
+  this._delete(`timetracking/business/${business_id}/time_entries/${time_entry_id}`, callback);
+}
+
+FreshBooks.prototype.loadProjectsForBiz = function(business_id, page, callback) {
```
@schuttsm: this is almost identical to `getProjects`; we should probably merge them into one method or if we really need two, have one call the other.

servant-42 commented on freshbooks.js:
```diff
@@ -10,84 +10,99 @@ function FreshBooks(access_token, refresh_token) {
+
+FreshBooks.prototype._isSuccessCode = function(res) {
+  var statusCode = res.statusCode.toString();
+  return statusCode[0] == '2';
```
@schuttsm: converting to a string & checking the first character works, but `res.statusCode >= 200 && res.statusCode < 300` is a bit clearer, more robust (not that `2` or `20` are ever going to happen) & more performant.

TBH, my preference would be to ignore the DRY and inline this where it's needed, since it's such a tiny bit of logic & is unlikely to ever need to change. If you really want to eliminate the DRY, this can be a helper function defined at the bottom of the file instead of a method. I realize this is OCD, but the code would be a little more obvious if the parameter was the response code (`isSuccess(res.statusCode)`), as opposed to the response object. I suppose one could argue that getting the `statusCode` property is a hint of DRY, but it's not logic & IMO passing `res.statusCode` reduces abstraction & makes things more obvious, so it's worth it).

servant-42 commented on tests.js:
```diff
@@ -1,44 +1,208 @@
+    firebase.initializeApp(config);
+    firebase.auth().signInWithEmailAndPassword(
+      process.env.FIREBASE_USERNAME,
+      process.env.FIREBASE_PASSWORD
```
@schuttsm: this make me nervous; can we switch to an auth token? This should probably be a new free-tier Firebase account (if it's not already).

schuttsm replied:
@prust Yes, this is a separate, free-tier firebase account.  The only data in it is a single "env_data" object with either a "code" or "refresh_token" property.  If the account becomes compromised, all they can do is screw up the refresh token so the build fails.

I'm open to switching to an auth token (seems a bit overkill, but better safe than sorry when it comes to security).  Did you want one generated with firebase-token-generator like in daycast?  It looks like that method is deprecated in the version of firebase we're using.

The new way is to generate a private key & connect with an admin account and then generate a token that way.  Again, I don't mind doing that, but storing a private key seems the same as storing a password.  What do you think?

servant-42 replied:
@schuttsm: I'm not familiar with the new Firebase (and I'm hoping to transition Daycast directly from the old one to Postgres, so I'm not super-interested in learning), so do what seems right to you. I don't want to fight the framework -- signing into a service w/ creds smelled like an antipattern to me, but if that's the typical/recommended way w/ the new Firebase, so be it.

servant-42 commented on package.json:
```diff
@@ -4,7 +4,7 @@
   "main": "freshbooks.js",
   "scripts": {
-    "test": "mocha tests.js"
+    "test": "./node_modules/.bin/mocha tests.js"
```
@schuttsm: the `./node_modules/.bin/` shouldn't be necessary, should it?

schuttsm replied:
If mocha is installed globally it works, but this works off of a simple "npm install"

I'm getting a "-bash: mocha: command not found" when I run it without the ./node_mo... on my machine

servant-42 commented on freshbooks.js:
```diff
@@ -10,84 +10,99 @@ function FreshBooks(access_token, refresh_token) {
-    });
-
-    var me = _.pick(me, required_props);
-    me.business_ids = business_ids;
```
@schuttsm: My intent (with `pick()` and `business_ids`) was to make the API flatter (easier to use) and to restrict our use of the API to the subset of properties/objects that are being tested. We should probably also allow access to underlying methods (perhaps underscore-prefixed?) that provide "raw" access to the API, but a lot of the value we're adding is providing a documented & tested stable subset of an alpha API.

Going a little further, I think we should add `active_business_ids`, which provides a little more value, filtering the business IDs to the ones with active subscriptions (since querying a business with a lapsed subscription results in an error). I'm not sure if we should also include `business_ids` or not -- perhaps we should, as it may make things a little less mysterious to an end user.

IOW, the more FreshBooks-specific logic, business rules, etc, that we can codify into this library (and out of our codebase) the better. It will make our app's codebase simpler and will provide value to others, so they don't have to learn the same things we've had to learn.

schuttsm replied:
@prust - that makes sense.

I was under the impression that this project was *only* intended to wrap tests around an api to let us know when it changed.

If you want to massage the api in a way that makes more logical sense, that's fine - I think I'd favor keeping the raw data available & add other methods for that.

prust reviewed (commented):
thanks @schuttsm! Review comments below.

Commit:
review changes

Commit:
review cleanup

Commit:
review cleanup

Commit:
review changes.  move to firebase admin account & token generation

Commit:
move firebase admin stuff to gen-token & use a token instead

Commit:
add an option for massaging the me api call or returning raw data (default raw=false).  Readd the token generation on each step because tokens only last 1 hr (no override)

Commit:
adjust tests to handle the travis ci env var limitations

Commit:
base64 encode private key before putting in env var, then decode before use

Commit:
cleanup

schuttsm commented:
@prust - ok, ready for another round of review.  I got the firebase token setup working, it was pretty easy to setup, once I figured out the private key needed to be base64 encoded before going into travis-ci's environment variable.

And I returned the "me" api call to be shaped as you had it before (with an optional "raw" flag to return the raw api data), except I moved the error handling to the travis tests as it seemed more natural there.  And I added the active_business_ids there too.

What do you think?  Should I transform the data from any other api calls?  Did you want me to start pulling this into daycast & centralizing the freshbooks api calls?

servant-42 commented on package.json:
```diff
@@ -23,13 +23,13 @@
+    "dotenv": "^4.0.0",
+    "firebase": "^4.12.1",
+    "firebase-admin": "^5.12.0",
+    "mocha": "^5.0.0",
```
@schuttsm: Shouldn't chai & mocha be devDependencies & the rest be dependencies?

prust reviewed (approved):
@schuttsm: review+ except for the question below re deps vs dev-deps (mocha & chai).

> I moved the error handling to the travis tests as it seemed more natural there

Sounds good.

> I returned the "me" api call to be shaped as you had it before (with an optional "raw" flag to return the raw api data)

We should probably call this good for now & cycle back to things that deliver value to the end-user, but eventually I would like to brainstorm more of a separate API for the raw data, as opposed to a flag. Maybe even exposing the lower-level `fb.get(path)` and `fb.post(path)`, etc, would be enough.

> Should I transform the data from any other api calls?

Probably not until you see a need for it...

> Did you want me to start pulling this into daycast & centralizing the freshbooks api calls?

We should pull it into daycast sooner, rather than later. But maybe it'd be good to cycle over to one or two user-facing tasks first, just to help a bit with momentum, before circling back & integrating this into Daycast.

Commit:
adjust package.json
